### PR TITLE
General underactuated base

### PR DESCRIPTION
**Pull request by gfadini**
This PR should be a solution to #807, I still have to properly test it on some robot whose base is not FloatingBase, but it doesn't change anything if it is

### CODE REVIEW

File: include/crocoddyl/multibody/actuations/floating-base.hpp [hunk 1]
```diff
@@ -25,13 +25,14 @@ class ActuationModelFloatingBaseTpl : public ActuationModelAbstractTpl<_Scalar>
   typedef StateMultibodyTpl<Scalar> StateMultibody;
   typedef typename MathBase::VectorXs VectorXs;
   typedef typename MathBase::MatrixXs MatrixXs;
+  int nvFirstJoint;
```
Comment (cmastalli) [inline include/crocoddyl/multibody/actuations/floating-base.hpp:28]: We do not need to create this object. I will provide a solution above

File: include/crocoddyl/multibody/actuations/floating-base.hpp [hunk 2]
```diff
@@ -25,13 +25,14 @@ class ActuationModelFloatingBaseTpl : public ActuationModelAbstractTpl<_Scalar>
...
 
-  explicit ActuationModelFloatingBaseTpl(boost::shared_ptr<StateMultibody> state) : Base(state, state->get_nv() - 6) {
+  explicit ActuationModelFloatingBaseTpl(boost::shared_ptr<StateMultibody> state) : Base(state, state->get_nv() - state->get_pinocchio()->joints[1].nv()) {
     pinocchio::JointModelFreeFlyerTpl<Scalar> ff_joint;
+    nvFirstJoint = state->get_pinocchio()->joints[1].nv();
```
Comment (cmastalli) [inline include/crocoddyl/multibody/actuations/floating-base.hpp:32]: Remove this line as mentioned above

File: include/crocoddyl/multibody/actuations/floating-base.hpp [hunk 3]
```diff
@@ -53,7 +54,7 @@ class ActuationModelFloatingBaseTpl : public ActuationModelAbstractTpl<_Scalar>
 
   virtual boost::shared_ptr<Data> createData() {
     boost::shared_ptr<Data> data = boost::allocate_shared<Data>(Eigen::aligned_allocator<Data>(), this);
-    data->dtau_du.diagonal(-6).fill((Scalar)1.);
+    data->dtau_du.diagonal(-nvFirstJoint).fill((Scalar)1.);
```
Comment (cmastalli) [inline include/crocoddyl/multibody/actuations/floating-base.hpp:57]: Change this code to:
```cpp
std::size_t fb_nv = state->get_pinocchio()->joints[1].nv();
data->dtau_du.diagonal(-fb_nv).fill((Scalar)1.);
```

Comment (gfadini) [reply include/crocoddyl/multibody/actuations/floating-base.hpp]: I tried that line before but it's not working, just not to add the variable I was not sure how to recover the state information I think I should use the ```state_``` there? Can you help me with this doubt?

**Comment by proyan** [comment]
Hi @gfadini 
The actuationfloatingbase is built in coherence with the StateMultibody, and both of them assume that only the first joint can be a free-flyer.
I think you should fix the constructor of StateMultibody as well to be more generic. Could you take care of that in this PR?

File: include/crocoddyl/multibody/actuations/floating-base.hpp (hunk 3 above)
Comment (cmastalli) [reply include/crocoddyl/multibody/actuations/floating-base.hpp]: Yes, that's correct.

File: include/crocoddyl/multibody/actuations/floating-base.hpp [hunk 4]
```diff
@@ -26,12 +26,11 @@ class ActuationModelFloatingBaseTpl : public ActuationModelAbstractTpl<_Scalar>
   typedef typename MathBase::VectorXs VectorXs;
   typedef typename MathBase::MatrixXs MatrixXs;
 
-  explicit ActuationModelFloatingBaseTpl(boost::shared_ptr<StateMultibody> state) : Base(state, state->get_nv() - 6) {
+  explicit ActuationModelFloatingBaseTpl(boost::shared_ptr<StateMultibody> state) : Base(state, state->get_nv() - state->get_pinocchio()->joints[1].nv()) {
     pinocchio::JointModelFreeFlyerTpl<Scalar> ff_joint;
     if (state->get_pinocchio()->joints[1].shortname() != ff_joint.shortname()) {
```
Comment (cmastalli) [inline include/crocoddyl/multibody/actuations/floating-base.hpp:31]: Let's remove this warming message

File: include/crocoddyl/multibody/actuations/floating-base.hpp [hunk 5]
```diff
@@ -50,10 +49,11 @@ class ActuationModelFloatingBaseTpl : public ActuationModelAbstractTpl<_Scalar>
     assert_pretty(data->dtau_dx == MatrixXs::Zero(state_->get_nv(), state_->get_ndx()), "dtau_dx has wrong value");
     assert_pretty(data->dtau_du == dtau_du_, "dtau_du has wrong value");
   };
-
+  boost::shared_ptr<StateMultibody> state = boost::dynamic_pointer_cast<StateMultibody> (state_);
```
Comment (cmastalli) [inline include/crocoddyl/multibody/actuations/floating-base.hpp:52]: Do not create this variable.
You can downcast it inside `createData`, i.e.
```cpp
virtual boost::shared_ptr<Data> createData() { 
  typedef StateMultibodyTpl<Scalar> StateMultibody;
  boost::shared_ptr<StateMultibody> state = boost::static_pointer_cast<StateMultibody>(state_);
  std::size_t fb_nv = state->get_pinocchio()->joints[1].nv();
...
}
```

File: include/crocoddyl/multibody/states/multibody.hxx [hunk 6]
```diff
@@ -16,24 +16,17 @@ template <typename Scalar>
...
-    nq0 = 7;
-    lb_.template head<7>() = -std::numeric_limits<Scalar>::infinity() * VectorXs::Ones(7);
-    ub_.template head<7>() = std::numeric_limits<Scalar>::infinity() * VectorXs::Ones(7);
-  } else if (model->joints[1].shortname() == "JointModelSphericalZYX") {
-    joint_type_ = Spherical;
-  }
+
+  // Define internally the free-flyer limits by removing the first joint DOFs
+
+  const std::size_t nq0 = model->joints[1].nq();
```
Comment (cmastalli) [inline include/crocoddyl/multibody/states/multibody.hxx:29]: Define the unbounded condition of the unactuated joint, i.e.
```cpp
    lb_.template head(nq0) = -std::numeric_limits<Scalar>::infinity() * VectorXs::Ones(nq0);
    ub_.template head(nq0) = std::numeric_limits<Scalar>::infinity() * VectorXs::Ones(nq0);
```

File: include/crocoddyl/multibody/states/multibody.hxx [hunk 7]
```diff
@@ -16,24 +16,17 @@ template <typename Scalar>
...
-    joint_type_ = FreeFlyer;
-    // Define internally the free-flyer limits
-    nq0 = 7;
-    lb_.template head<7>() = -std::numeric_limits<Scalar>::infinity() * VectorXs::Ones(7);
-    ub_.template head<7>() = std::numeric_limits<Scalar>::infinity() * VectorXs::Ones(7);
-  } else if (model->joints[1].shortname() == "JointModelSphericalZYX") {
-    joint_type_ = Spherical;
-  }
+
+  // Define internally the free-flyer limits by removing the first joint DOFs
```
Comment (cmastalli) [inline include/crocoddyl/multibody/states/multibody.hxx:27]: // Define internally the limits of the unactuated joint

**Comment by gfadini** [comment]
You're welcome, thanks for the tips in modifying the code!

File: include/crocoddyl/multibody/states/multibody.hxx (hunk 6 above)
Comment (proyan) [reply include/crocoddyl/multibody/states/multibody.hxx:29]: You should not use the `template` keyword, when the call is not calling a template function. `lb_.head()` does not have a template argument, and this code should produce errors. I'm surprised that it passed CI!

Comment (cmastalli) [reply include/crocoddyl/multibody/states/multibody.hxx:29]: Do you think this fails only with codegen compilation?

**Pull request merged**